aquadoggo/src/materializer/tasks/reduce.rs

```diff
@@ -1,17 +1,242 @@
 // SPDX-License-Identifier: AGPL-3.0-or-later
 
 use p2panda_rs::document::DocumentBuilder;
+use p2panda_rs::storage_provider::traits::OperationStore;
 
 use crate::context::Context;
-use crate::materializer::worker::{TaskError, TaskResult};
+use crate::db::traits::DocumentStore;
+use crate::materializer::worker::{Task, TaskError, TaskResult};
 use crate::materializer::TaskInput;
 
-pub async fn reduce_task(_context: Context, _input: TaskInput) -> TaskResult<TaskInput> {
-    // @TODO: Load operations from database
-    let _document = DocumentBuilder::new(vec![])
-        .build()
-        .map_err(|_| TaskError::Failure)?;
+pub async fn reduce_task(context: Context, input: TaskInput) -> TaskResult<TaskInput> {
+    let document_id = match (&input.document_id, &input.document_view_id) {
+        (Some(document_id), None) => Ok(document_id.to_owned()),
+        // TODO: Alt approach: we could have a `get_operations_by_document_view_id()` in `OperationStore`, or
+        // could we even do this with some fancy recursive SQL query? We might need the `previous_operations`
+        // table back for that.
+        (None, Some(document_view_id)) => {
+            let operation_id = document_view_id.clone().into_iter().next().unwrap();
+            match context
+                .store
+                .get_document_by_operation_id(operation_id)
+                .await
+                .map_err(|_| TaskError::Critical)? // We only get an error here on a critical database error.
+            {
+                Some(document_id) => Ok(document_id),
+                None => Err(TaskError::Failure),
+            }
+        }
+        (_, _) => Err(TaskError::Failure),
+    }?;
 
-    // @TODO: Store document view in database
-    Ok(None)
+    let operations = context
+        .store
+        .get_operations_by_document_id(&document_id)
+        .await
+        .map_err(|_| TaskError::Critical)?
+        .into_iter()
+        // TODO: we can avoid this conversion if we do https://github.com/p2panda/p2panda/issues/320
+        .map(|op| op.into())
+        .collect();
+
+    let document_view_id = match &input.document_view_id {
+        Some(document_view_id) => {
+            // TODO: If we are resolving a document_view_id, but we are missing operations from it's graph,
+            // what do we want to return?
+            let document = DocumentBuilder::new(operations)
+                .build_to_view_id(Some(document_view_id.to_owned()))
+                .map_err(|_| TaskError::Critical)?;
+
+            if document.is_deleted() {
+                return Ok(None);
```

Comment:
That's an interesting case which we haven't looked too much yet: What happens when a document gets deleted?

I haven't thought it fully through yet but I can imagine we actually want to dispatch dependency tasks for it as well as this might again have effects further down the line. For example: a deleted schema definition document will probably remove the item in the schema service (or set it to deprecated etc., we haven't decided yet).

It's good to keep it simple for now but maybe we should make a TODO / issue for that.